BugFix: make optional part of package exporter UI expand if there is space

SlySven opened the pull request:
This will close #5352.

The issue is that several `QLayouts` have been included on their own instead of being assigned to a `QWidget` or other container widget. The fix is to "morph" them into `QWidgets` which makes the later take on that layout. In doing so I realised that where there were nested layouts (one inside the other) there was some duplication of left/right margins so that some of the displayed widgets were not as wide as some others; this was improved by setting some of the left/right margins to zero.

It also corrects the spelling of a `QPushButton` pointer,
from: `pushBotton_packageLocation`
to: `pushButton_packageLocation`

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: make optional part of package exporter UI expand if there is space

This will close #5352.

The issue is that several `QLayouts` have been included on their own
instead of being assigned to a `QWidget` or other container widget. The fix
is to "morph" them into `QWidgets` which makes the later take on that
layout. In doing so I realised that where there were nested layouts (one
inside the other) there was some duplication of left/right margins so that
some of the displayed widgets were not as wide as some others; this was
improved by setting some of the left/right margins to zero.

It also corrects the spelling of a `QPushButton` pointer,
from: `pushBotton_packageLocation`
to: `pushButton_packageLocation`

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Hmm, before on Linux 2K display:

![image](https://user-images.githubusercontent.com/110988/129435550-6a2a5fa4-8095-46c5-9f76-d4736652775d.png)


Now -

![image](https://user-images.githubusercontent.com/110988/129435537-29c75de7-9c12-4efb-95ff-df9d0c93b5a8.png)

The export buttons don't fit anymore at the bottom.

vadi2 commented:
Not able to test on macOS at the moment, our machine is out of action - working on that separately.

vadi2 commented:
macOS access restored. This is how it looks like -

![image](https://user-images.githubusercontent.com/110988/129440393-3078c990-d316-4000-bb1e-d3e103b93803.png)

SlySven committed:
Revise: further tweaks to make some elements expand horizontally

Also fix another spelling mistake in `pushBotton_openInfos`
==> `pushButton_openInfos` .

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> https://github.com/Mudlet/Mudlet/pull/5372#issuecomment-898830911

It seems like the original mis-applied/used `QLayout`s did not apply the default margins/spacings that are now present in the "fixed" version. Let me see if I can trim those things a little in the latter so that the widgets are drawn a bit closer together...

vadi2 commented:
It's still being difficult.

![image](https://user-images.githubusercontent.com/110988/129471223-94e8ef69-b1f6-4ddc-a464-482fa7bc8e5a.png)

SlySven commented:
> It's still being difficult.

That is weird. At the moment I am away from home (visiting parent for a long weekend) and on my Laptop with its 1280x800 NON-HiRes display the expanding to fit is working:
![Screenshot_20210815_162806](https://user-images.githubusercontent.com/6163092/129483707-fb2238ec-5fcc-4577-a487-654a8c5a1266.png)

Note that I have shaved a little too much away from the PushButton to show/hide the optional detail from the first line of that detail (the "Author" one) so another tweak is needed anyhow.

I guess this will have to wait until I get home so that I can see what it looks like on my 4K TV/Display but something aint working right...

SlySven commented:
I was going to point out that the `dlgPackageExporter.ui` file is one that is compiled into the `mudlet.qrc` resource file and perhaps a second build might be needed for any changes made to the source file gets into the resource file that the application uses - but then I saw that it is not the case and that file is `uic`ed into the `ui_dlgPackageExporter.h` file `#includ`ed into the `dlgPackageExporter.cpp` file so updates *should* get in the first time a revised version is built.

Unless I am mistaken. Try to edit and then undo the edit to `dlgPackageExporter.cpp` (or delete the `ui_dlgPackageExporter.h` file - not sure where it ends up) so as to regenerate the `dlgPackageExporter.o` file (or whatever the CMake build system produces as object files nowadays) and see if that produces any changes...?

SlySven committed:
Revise: more vertical spacing tweaks in elements in packageExporter dialog

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Still no luck -

![image](https://user-images.githubusercontent.com/110988/130582638-518cdbb6-81b6-4f85-a320-3c48e2234bf5.png)

Maybe there is something macOS specific here that we have to do with the description area?

SlySven commented:
> Maybe there is something macOS specific here that we have to do with the description area?

Which Qt version is that on BTW? Given that it is going to be a Qt 5.x version I doubt we are going to see a fix any time soon if it happens to be a MacOs ~~defect~~ *feature*...

:frowning_face:

vadi2 commented:
https://github.com/Mudlet/Mudlet/blob/development/.github/workflows/build-mudlet.yml#L36-L40

SlySven commented:
Humm, web-browsing led me to https://forum.qt.io/topic/52888/qlineedit-doesn-t-stretch-to-occupy-full-space-in-the-layout which in turn suggested the change that I was going to make in the UI file:
```xml
           </property>
           <layout class="QFormLayout" name="formLayout_metadata">
+           <property name="sizeConstraint">
+            <enum>QLayout::SetNoConstraint</enum>
+           </property>
            <property name="leftMargin">
             <number>0</number>
            </property>
 ```
this led to: https://doc.qt.io/qt-5/qlayout.html#public-types and then it became clear that the problems are because someone decided to use a `QFormLayout` for that part of the data - and the behaviour of that specialized layout {two columns, left column containing `QLabels`} **has platform specific variations** - see https://doc.qt.io/qt-5/qformlayout.html - in particular for the `QMacStyle`:
> Style based on the macOS Aqua guidelines. Labels are right-aligned, the fields don't grow beyond their size hint, and the form is horizontally centered.

So I think if I change that part to use a `QGridLayout` with two columns all platforms will behave the same.

SlySven committed:
Revise: finally nail the issue

It is because the problematic section of the form uses a `QFormLayout`
which is styled differently on MacOs (`QMacStyle`) compared to the styling
traditionally used for other platforms (`QCommonStyle`) - though modern
KDE desktops will use something a bit in between (`QPlastiqueStyle`) though
that still expands the widgets on the right side of a `QFormLayout`.

It should be possible to get the same behaviour on ALL platforms by
dropping back to a `QGridLayout`.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
/refresh links

vadi2 reviewed (approved):
Yep nice, this works!

Merging so it makes it into 4.13

SlySven commented:
Ah, you didn't edit the combined commit message when you squashed all the commits together @vadi2 (which I would have done) fortunately there isn't anything erroneous or misleading in those messages for this PR - it is just a bit untidy - but beware of that when there is more than one commit involved (particularly in some of my more *epic* PRs!)
